miden-lib/asm/miden/note.masm

```diff
@@ -143,4 +143,435 @@ export.get_serial_number
+
+        # push latch variable onto the stack; this will be the control for the cdrop instruction
+        push.0
+        # => [latch = 0, e_1, 0/1, e_2, e_3, d_0, d_1, d_2, 0/1, d_3, A', drop_counter]
+
+        # get the flag whether the drop counter is equal 1
+        dup.14 eq.1
+        # => [drop_counter == 1, latch = 0, e_1, 0/1, e_2, e_3, d_0, d_1, d_2, 0/1, d_3, A', drop_counter]
+
+        # update the latch: if drop_counter == 1, latch will become 1
+        or
+        # => [latch', e_1, 0/1, e_2, e_3, d_0, d_1, d_2, 0/1, d_3, A', drop_counter]
+
+        # save the latch value 
+        dup movdn.14
+        # => [latch', e_1, 0/1, e_2, e_3, d_0, d_1, d_2, 0/1, d_3, A', latch', drop_counter]
+
+        # if latch == 1, drop 0/1; otherwise drop e_1
+        cdrop
+        # => [e_1_or_0/1, e_2, e_3, d_0, d_1, d_2, 0/1, d_3, A', latch', drop_counter]
+
+        # move the calculated value down the stack 
+        movdn.6
+        # => [e_2, e_3, d_0, d_1, d_2, 0/1, e_1_or_0/1, d_3, A', latch', drop_counter]
+
+        ### 2nd value ########################################################
+
+        # repeat the above process but now compare drop_counter to 2
+        dup.13 eq.3 swap
+        movup.13 dup.14 eq.2 or
+        dup movdn.14
+        cdrop movdn.6
+        # => [e_3, d_0, d_1, d_2, 0/1, e_1_or_0/1, e_2_or_0/1, d_3, A', latch', drop_counter]
+
+        ### 3rd value ########################################################
+
+        # repeat the above process but now compare drop_counter to 3
+        dup.13 eq.4 swap
+        movup.13 dup.14 eq.3 or
+        dup movdn.14
+        cdrop movdn.6
+        # => [d_0, d_1, d_2, 0/1, e_1_or_0/1, e_2_or_0/1, e_3_or_0/1, d_3, A', latch', drop_counter]
+
+        ### 4th value ########################################################
+
+        # repeat the above process but now compare drop_counter to 4
+        dup.13 eq.5 swap
+        movup.13 dup.14 eq.4 or
+        dup movdn.14
+        cdrop movdn.6
+        # => [d_1, d_2, 0/1, e_1_or_0/1, e_2_or_0/1, e_3_or_0/1, d_0_or_0/1, d_3, A', latch', drop_counter]
+
+        ### 5th value ########################################################
+
+        # repeat the above process but now compare drop_counter to 5
+        dup.13 eq.6 swap
+        movup.13 dup.14 eq.5 or
+        dup movdn.14
+        cdrop movdn.6
+        # => [d_2, 0/1, e_1_or_0/1, e_2_or_0/1, e_3_or_0/1, d_0_or_0/1, d_1_or_0/1, d_3, A', latch', drop_counter]
+
+        ### 6th value ########################################################
+
+        # repeat the above process but now compare drop_counter to 6
+        dup.13 eq.7 swap
+        movup.13 dup.14 eq.6 or
+        cdrop movdn.6
+        # => [0/1, e_1_or_0/1, e_2_or_0/1, e_3_or_0/1, d_0_or_0/1, d_1_or_0/1, d_2_or_0/1, d_3, A', drop_counter]
+        # or in other words
+        # => [C, B, A', drop_counter, ... ]
+
+        # drop the drop_counter
+        movup.12 drop
```

Comment:
if we use `movdn` instead of `dup` in the block above, we may be able to drop the `drop_counter` naturally and then we won't need these instructions here.